### PR TITLE
JPA-Map storage might loose writes due to missing locking mechanism

### LINKED ISSUE #9411: JPA-Map storage might loose writes due to missing locking mechanism
### Describe the bug

The JPA implementation for the map storage provider doesn't specify a locking strategy, therefore writes might get lost: When Hibernate writes the contents back to the database, the contents might have been updated by another concurrent request already.
In an extreme case this could lead to some column being updated while others aren't, leaving the database in an inconsistent state. 

### Version

17.0.0-SNAPSHOT

### Expected behavior

When persisting data in the JPA map storage provider 
succeeds, the contents should be stored consistently in all related tables (main entity table plus additional tables for searching). 

When a concurrent transaction modifies the data, the transaction should fail with an exception. 

### Actual behavior

When a concurrent write occurred, Hibernate will overwrite them without noticing some other transaction has modified the data in between. As Hibernate will write only content to the database it considered changed, this could lead to a partial overwrite of the concurrent transaction, leading to inconsistent data.

### How to Reproduce?

Looking at the code, no JPA entity of the map storage has a version column. The option of `optimistic-lock="all"` is not used AFAIK.

While the old JPA storage uses PESSIMISTIC_WRITE in several places (usually when deleting data), there is no such code in the new JPA storage provider.

### Anything else?

**Inside of the scope of this issue:** The old storage uses PESSIMISTIC_WRITE in several places, find out the rationale why it did so, and if this should apply to the new JPA map storage as well.

**Outside of the scope of this issue:** As the admin API of Keycloak doesn't provide a mechanism for optimistic locking or identification of concurrent updates by multiple clients (there is not version ID or similar in the API), lost writes when accessing the REST API might still occur. Adding a mechanism for that in the API of map storage and other APIs is out of scope here. This is only about keeping the JPA storage consistent.

### PR DESCRIPTION
**Pull request by vramik**
Closes #9411

### CODE REVIEW

**Review by ahus1 — DISMISSED** [review]
Hi @vramik - I reviewed the changes and they look good to me. I double-checked if we would need a version on the child table as well, but the docs say no (see https://vladmihalcea.com/hibernate-collections-optimistic-locking/ in "Collection versioning"). 

Therefore, for clients this is good to go. 

I wonder how you want to proceed for the version on role: separate pull request, add it to the other PR, or wait for the other PR to merged first, then rebase this and add the other version. Your decision :-)

**Comment by vramik** [comment]
Thank you @ahus1 

> I wonder how you want to proceed for the version on role: separate pull request, add it to the other PR, or wait for the other PR to merged first, then rebase this and add the other version. Your decision :-)

I think the best would be to merge this one first and in the meantime I'll add it to the roles PR.

File: model/map-jpa/src/main/resources/META-INF/clients/jpa-clients-changelog-1.xml [hunk 1]
```diff
@@ -29,6 +29,7 @@ limitations under the License.
             <column name="id" type="UUID">
                 <constraints primaryKey="true" nullable="false"/>
             </column>
+            <column name="version" type="INTEGER"/>
```
Comment (hmlnarik) [inline model/map-jpa/src/main/resources/META-INF/clients/jpa-clients-changelog-1.xml:32]: Should this column be not null with default value 0?

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java [hunk 2]
```diff
@@ -54,6 +55,10 @@
     @Column
     private UUID id;
 
+    //used for implicit optimistic locking
+    @Version
+    private int version;
```
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:60]: Should there be a `@Column` annotation?

Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:60]: The `@Column` annotation is optional from a technical point of view as long as you don't use it to change the database field name, type or to add constraints. Basically every time we use the "plain" annotation without any parameters we could leave it out.

File: model/map-jpa/src/main/resources/META-INF/clients/jpa-clients-changelog-1.xml (hunk 1 above)
Comment (ahus1) [reply model/map-jpa/src/main/resources/META-INF/clients/jpa-clients-changelog-1.xml:32]: As this is the initial setup of the table with no data in it, the default value would be optional. 
+1 for "not null". 

If we want consistency on this for version columns, there should be a unit test for that (but that should be a separate GitHub issue).

**Review by hmlnarik — APPROVED** [review]

**Pull request merged**